File: src/lib/watermark/hasselblad.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  cleanDisplayText,
  formatTakenAt,
  getPreviewFontFamily,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  paramsLine,
  type ShowcasePhoto,
} from "./renderer-utils";

const HASSY_BG = "#0a0a0a";
const HASSY_ORANGE = "#ff8a00";
const HASSY_INK = "#f5f5f5";
const HASSY_MUTED = "#9a958c";

/**
 * Hasselblad style. Pure black canvas, photo near-fullbleed with bold
 * "HASSELBLAD" wordmark below it (orange), and EXIF aligned right. The
 * watermark stays minimalist — only camera/params, no lens or extras.
 */
export function drawHasselbladPreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  const sideMargin = canvasW * 0.04;
  const topMargin = canvasH * 0.04;
  const captionH = watermarkActive ? Math.max(96, canvasH * 0.14) : 0;

  const photoArea = {
    x: sideMargin,
    y: topMargin,
    w: canvasW - sideMargin * 2,
    h: canvasH - topMargin - captionH - canvasH * 0.04,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = HASSY_BG;
  ctx.fillRect(0, 0, canvasW, canvasH);
  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const captionTop = placed.y + placed.h + canvasH * 0.025;

  // Left: HASSELBLAD wordmark in orange — wide letterspacing for the premium feel.
  const wordSize = Math.max(20, canvasW * 0.034);
  ctx.fillStyle = HASSY_ORANGE;
  ctx.textBaseline = "alphabetic";
  ctx.textAlign = "left";
  ctx.font = `700 ${wordSize}px ${fontFamily}`;
  drawSpacedText(ctx, "HASSELBLAD", placed.x, captionTop + wordSize * 0.9, wordSize * 0.18);

  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const lens = config.showLens ? cleanDisplayText(photo.exif.lens) : "";
  const subline = [camera, lens].filter(Boolean).join("  ·  ");
  if (subline) {
    ctx.fillStyle = HASSY_MUTED;
    ctx.font = `400 ${Math.max(11, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(subline, placed.x, captionTop + wordSize * 0.9 + Math.max(20, frameParams.fontSize * 1.6));
  }

  // Right: params + date.
  const rightX = placed.x + placed.w;
  const params = config.showParams ? paramsLine(photo.exif, "  /  ") : "";
  const date = config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "";

  ctx.textAlign = "right";
  if (params) {
    ctx.fillStyle = HASSY_INK;
    ctx.font = `500 ${Math.max(13, frameParams.fontSize * 1.2)}px ${fontFamily}`;
    ctx.fillText(params, rightX, captionTop + wordSize * 0.9);
  }
  if (date) {
    ctx.fillStyle = HASSY_MUTED;
    ctx.font = `400 ${Math.max(11, frameParams.fontSize)}px ${fontFamily}`;
    ctx.fillText(date, rightX, captionTop + wordSize * 0.9 + Math.max(20, frameParams.fontSize * 1.6));
  }
}

function drawSpacedText(
  ctx: CanvasRenderingContext2D,
  text: string,
  x: number,
  y: number,
  spacing: number,
) {
  let cx = x;
  for (const ch of text) {
    ctx.fillText(ch, cx, y);
    cx += ctx.measureText(ch).width + spacing;
  }
}

File: src-tauri/src/render/templates/hasselblad.rs
use image::{imageops::overlay, DynamicImage, Rgba, RgbaImage};

use crate::render::classic_bottom::{
    build_params_line, clean_display_text, compute_canvas_size, fit_photo_in_area, format_camera,
    format_date, resize_photo, ExportExif, ExportFrameParams, ExportTemplateConfig,
};
use crate::render::layout_spec::watermark_layout_spec;
use crate::render::text::TextRenderer;

const HASSY_BG: Rgba<u8> = Rgba([10, 10, 10, 255]);
const HASSY_ORANGE: Rgba<u8> = Rgba([255, 138, 0, 255]);
const HASSY_INK: Rgba<u8> = Rgba([245, 245, 245, 255]);
const HASSY_MUTED: Rgba<u8> = Rgba([154, 149, 140, 255]);

pub(crate) fn compose_hasselblad(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let side_margin = (canvas_w as f32 * 0.04).round() as u32;
    let top_margin = (canvas_h as f32 * 0.04).round() as u32;
    let bottom_pad = (canvas_h as f32 * 0.04).round() as u32;
    let caption_h = if config.show_watermark {
        ((canvas_h as f32 * 0.14).round() as u32).max((96.0 * rs) as u32)
    } else {
        0
    };

    let area_w = canvas_w.saturating_sub(side_margin * 2);
    let area_h = canvas_h.saturating_sub(top_margin + caption_h + bottom_pad);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let photo_x = (side_margin + ox) as i64;
    let photo_y = (top_margin + oy) as i64;
    let placed_x = side_margin + ox;
    let placed_y = top_margin + oy;

    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, HASSY_BG);

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, photo_x, photo_y);

    if !config.show_watermark {
        return canvas;
    }

    let Some(rend) = text else { return canvas; };

    let spec = watermark_layout_spec();
    let word_pt = ((canvas_w as f32 * 0.034).max(20.0)) * rs;
    let detail_pt = (frame.font_size as f32 * spec.secondary_font_scale).max(11.0 * rs);
    let caption_top = placed_y as f32 + ph as f32 + (canvas_h as f32 * 0.025).round();
    let word_baseline = caption_top + word_pt * 0.9;

    // HASSELBLAD with letter spacing
    let spacing = word_pt * 0.18;
    draw_spaced_text(&mut canvas, rend, "HASSELBLAD", placed_x as f32, word_baseline - rend.ascent(word_pt, true) + rend.line_height(word_pt, true) * 0.1, word_pt, spacing, HASSY_ORANGE);

    // Camera + lens subline
    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let lens = if config.show_lens { clean_display_text(&exif.lens) } else { String::new() };
    let subline: String = [camera, lens].into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join("  \u{00B7}  ");
    if !subline.is_empty() {
        let sub_y = word_baseline + (detail_pt * 1.6).max(20.0 * rs);
        rend.draw(&mut canvas, &subline, placed_x as f32, sub_y - rend.line_height(detail_pt, false), detail_pt, false, HASSY_MUTED);
    }

    // Right: params + date
    let right_x = placed_x + pw;
    let params_pt = (frame.font_size as f32 * spec.secondary_font_scale * 1.2).max(13.0 * rs);
    let params = if config.show_params { build_params_line(exif, "  /  ") } else { String::new() };
    let date = if config.show_date { format_date(&exif.taken_at, &config.date_format) } else { String::new() };

    if !params.is_empty() {
        let pw_meas = rend.measure(&params, params_pt, false);
        let py = word_baseline - rend.line_height(params_pt, false);
        rend.draw(&mut canvas, &params, right_x as f32 - pw_meas, py, params_pt, false, HASSY_INK);
    }
    if !date.is_empty() {
        let dw = rend.measure(&date, detail_pt, false);
        let dy = word_baseline + (detail_pt * 1.6).max(20.0 * rs) - rend.line_height(detail_pt, false);
        rend.draw(&mut canvas, &date, right_x as f32 - dw, dy, detail_pt, false, HASSY_MUTED);
    }

    canvas
}

fn draw_spaced_text(
    canvas: &mut RgbaImage,
    rend: &TextRenderer,
    text: &str,
    mut x: f32,
    y: f32,
    pt: f32,
    spacing: f32,
    color: Rgba<u8>,
) {
    for ch in text.chars() {
        let s = ch.to_string();
        rend.draw(canvas, &s, x, y, pt, true, color);
        x += rend.measure(&s, pt, true) + spacing;
    }
}
